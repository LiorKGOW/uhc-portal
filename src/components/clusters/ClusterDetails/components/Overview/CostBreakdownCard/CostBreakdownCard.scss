.ocm--cost-breakdown-card {

  &__body {
    display: flex;
    flex-direction: column;

    .ocm--cost-icon {
      height: 75px;
      margin-bottom: 0;
    }

    .ocm--cost-title {
      .pf-c-title {
        display: inline;
      }
    }

    .ocm--cost-total {
      font-weight: var(--pf-global--FontWeight--light);
    }

    .pf-c-empty-state {
      padding: 0;
    }
  }
}
