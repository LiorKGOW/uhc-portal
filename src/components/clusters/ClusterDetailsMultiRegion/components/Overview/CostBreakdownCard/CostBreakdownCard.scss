.ocm--cost-breakdown-card {
  &__body {
    display: flex;
    flex-direction: column;

    .ocm--cost-icon {
      height: 75px;
      margin-bottom: 0;
    }

    .ocm--cost-title {
      .pf-v5-c-title {
        display: inline;
      }
    }

    .ocm--cost-total {
      font-weight: var(--pf-v5-global--FontWeight--normal);
    }

    .pf-v5-c-empty-state {
      padding: 0;
    }
  }
}
