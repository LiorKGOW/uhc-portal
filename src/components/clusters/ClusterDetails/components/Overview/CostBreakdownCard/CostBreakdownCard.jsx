import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Card, CardBody, CardTitle, Title } from '@patternfly/react-core';

import CostEmptyState from '../../../../../overview/CostCard/CostEmptyState';
import CostBreakdownSummary from './CostBreakdownSummary';

import './CostBreakdownCard.scss';

class CostBreakdownCard extends Component {
  componentDidMount() {
    this.refresh();
  }

  refresh = () => {
    const { clusterId, getReport, getSources } = this.props;

    // Filter example: cluster = ['a94ea9bc-9e4f-4b91-89c2-c7099ec08427']
    getReport({
      filter: {
        cluster: clusterId,
      },
    });
    getSources({ type: 'OCP' });
  };

  render() {
    const { report, sources } = this.props;
    const hasSources = sources && sources.meta && sources.meta.count !== 0;

    return (
      <Card className="ocm--cost-breakdown-card">
        <CardTitle>
          <Title size="lg" headingLevel="h2">
            Cost breakdown
          </Title>
        </CardTitle>
        <CardBody className="ocm--cost-breakdown-card__body">
          {!hasSources && sources.fulfilled ? (
            <CostEmptyState />
          ) : (
            <CostBreakdownSummary report={report} />
          )}
        </CardBody>
      </Card>
    );
  }
}

CostBreakdownCard.propTypes = {
  clusterId: PropTypes.string,
  getReport: PropTypes.func.isRequired,
  getSources: PropTypes.func.isRequired,
  report: PropTypes.shape({
    data: PropTypes.array,
    meta: PropTypes.object,
    pending: PropTypes.bool,
    fulfilled: PropTypes.bool,
  }).isRequired,
  sources: PropTypes.shape({
    meta: PropTypes.object,
    pending: PropTypes.bool,
    fulfilled: PropTypes.bool,
  }).isRequired,
};

export default CostBreakdownCard;
