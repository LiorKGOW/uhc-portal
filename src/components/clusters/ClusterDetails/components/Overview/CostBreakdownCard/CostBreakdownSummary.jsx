import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Skeleton from '@redhat-cloud-services/frontend-components/Skeleton';
import { Title } from '@patternfly/react-core';
import { ChartLabel, ChartLegend, ChartPie } from '@patternfly/react-charts';

import './CostBreakdownCard.scss';

const CHART_HEIGHT = 185;
const CHART_WIDTH = 350;

class CostBreakdownSummary extends Component {
  formatCurrency = (value = 0, units = 'USD') =>
    value.toLocaleString('en', {
      style: 'currency',
      currency: units,
      minimumFractionDigits: 2,
      maximumFractionDigits: 2,
    });

  getChart = () => {
    const { report } = this.props;

    const hasCost = report && report.meta && report.meta.total && report.meta.total.cost;
    const hasMarkup = hasCost && report.meta.total.cost.markup;
    const hasRaw = hasCost && report.meta.total.cost.raw;
    const hasUsage = hasCost && report.meta.total.cost.usage;

    const markupUnits = hasMarkup ? report.meta.total.cost.markup.units : 'USD';
    const rawUnits = hasRaw ? report.meta.total.cost.raw.units : 'USD';
    const usageUnits = hasUsage ? report.meta.total.cost.usage.units : 'USD';

    const markupValue = hasMarkup ? report.meta.total.cost.markup.value : 0;
    const rawValue = hasRaw ? report.meta.total.cost.raw.value : 0;
    const usageValue = hasUsage ? report.meta.total.cost.usage.value : 0;

    const markup = this.formatCurrency(
      hasMarkup ? report.meta.total.cost.markup.value : 0,
      markupUnits,
    );
    const raw = this.formatCurrency(hasRaw ? report.meta.total.cost.raw.value : 0, rawUnits);
    const usage = this.formatCurrency(
      hasUsage ? report.meta.total.cost.usage.value : 0,
      usageUnits,
    );

    const markupLabel = 'Markup';
    const rawLabel = 'Raw cost';
    const usageLabel = 'Usage cost';

    return (
      <ChartPie
        ariaDesc="Cost breakdown"
        constrainToVisibleArea
        data={[
          { x: rawLabel, y: rawValue, units: rawUnits },
          { x: markupLabel, y: markupValue, units: markupUnits },
          { x: usageLabel, y: usageValue, units: usageUnits },
        ]}
        height={CHART_HEIGHT}
        labels={({ datum }) => `${datum.x}: ${this.formatCurrency(datum.y, datum.units)}`}
        legendComponent={this.getLegend([raw, markup, usage])}
        legendData={[
          {
            name: rawLabel,
          },
          {
            name: markupLabel,
          },
          {
            name: usageLabel,
          },
        ]}
        legendOrientation="vertical"
        legendPosition="right"
        padding={{
          bottom: 30,
          left: 0,
          right: 225, // Adjusted to accommodate legend
          top: 30,
        }}
        legendAllowWrap
        width={CHART_WIDTH}
      />
    );
  };

  // Override legend label layout
  getLegendLabel =
    () =>
    ({ values, ...props }) => (
      <ChartLabel
        {...props}
        style={[{ fontWeight: 600 }, {}]}
        text={[values[props.index], props.text]}
      />
    );

  getLegend = (values) => {
    const LegendLabel = this.getLegendLabel();
    return (
      <ChartLegend
        gutter={25}
        itemsPerRow={2}
        labelComponent={<LegendLabel dy={10} lineHeight={1.5} values={values} />}
        rowGutter={20}
      />
    );
  };

  getTotal = () => {
    const { report } = this.props;

    const hasTotal =
      report &&
      report.meta &&
      report.meta.total &&
      report.meta.total.cost &&
      report.meta.total.cost.total;
    const total = hasTotal ? report.meta.total.cost.total.value : 0;
    const units = hasTotal ? report.meta.total.cost.total.units : 'USD';

    return this.formatCurrency(total, units);
  };

  render() {
    const { report } = this.props;

    if (!report.fulfilled) {
      return <Skeleton size="md" />;
    }

    return (
      <>
        <Title className="ocm--cost-title" size="md" headingLevel="h2">
          Total cost
          <span className="ocm--cost-total"> {this.getTotal()}</span>
        </Title>
        <div style={{ maxHeight: CHART_HEIGHT, maxWidth: CHART_WIDTH }}>{this.getChart()}</div>
      </>
    );
  }
}

CostBreakdownSummary.propTypes = {
  report: PropTypes.shape({
    data: PropTypes.array,
    meta: PropTypes.object,
    pending: PropTypes.bool,
    fulfilled: PropTypes.bool,
  }).isRequired,
  values: PropTypes.array,
};

export default CostBreakdownSummary;
