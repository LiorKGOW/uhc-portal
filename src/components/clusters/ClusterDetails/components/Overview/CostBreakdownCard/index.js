import { connect } from 'react-redux';
import CostBreakdownCard from './CostBreakdownCard';
import { getReport, getSources } from '../../../../../../redux/actions/costActions';

const mapDispatchToProps = {
  getReport,
  getSources,
};

const mapStateToProps = state => ({
  report: state.cost.report,
  sources: state.cost.sources,
});

// const mapDispatchToProps = dispatch => ({
//   getReport: params => dispatch(getReport(params)),
//   getSources: params => dispatch(getSources(params)),
// });

export default connect(mapStateToProps, mapDispatchToProps)(CostBreakdownCard);
