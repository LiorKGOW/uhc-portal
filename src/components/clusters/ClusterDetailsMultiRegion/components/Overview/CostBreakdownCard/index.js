import { connect } from 'react-redux';

import { getReport, getSources } from '../../../../../../redux/actions/costActions';

import CostBreakdownCard from './CostBreakdownCard';

const mapDispatchToProps = {
  getReport,
  getSources,
};

const mapStateToProps = (state) => ({
  report: state.cost.report,
  sources: state.cost.sources,
});

export default connect(mapStateToProps, mapDispatchToProps)(CostBreakdownCard);
