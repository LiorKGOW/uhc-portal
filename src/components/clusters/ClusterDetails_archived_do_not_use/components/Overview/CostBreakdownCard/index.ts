import CostBreakdownCard from './CostBreakdownCard';

export default CostBreakdownCard;
