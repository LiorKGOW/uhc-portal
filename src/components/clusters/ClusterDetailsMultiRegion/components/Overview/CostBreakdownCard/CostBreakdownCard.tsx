import React, { useEffect, useMemo } from 'react';
import { useDispatch } from 'react-redux';

import { Card, CardBody, CardTitle, Title } from '@patternfly/react-core';

import { Report } from '~/components/dashboard/CostCard/models/Report';
import { getReport, getSources } from '~/redux/actions/costActions';
import { useGlobalState } from '~/redux/hooks';

import CostEmptyState from '../../../../../dashboard/CostCard/CostEmptyState';

import CostBreakdownSummary from './CostBreakdownSummary';

import './CostBreakdownCard.scss';

type CostBreakdownCardProps = {
  clusterId?: string;
};
const CostBreakdownCard = ({ clusterId }: CostBreakdownCardProps) => {
  const dispatch = useDispatch();
  const { report, sources } = useGlobalState((state) => state.cost);

  useEffect(() => {
    dispatch(
      getReport({
        filter: {
          cluster: clusterId,
        },
      }),
    );
    dispatch(getSources({ type: 'OCP' }));
  }, [clusterId, dispatch]);

  const hasSources = useMemo(
    () => sources?.meta?.count !== undefined && sources?.meta?.count !== 0,
    [sources?.meta?.count],
  );

  return (
    <Card className="ocm--cost-breakdown-card">
      <CardTitle>
        <Title size="lg" headingLevel="h2">
          Cost breakdown
        </Title>
      </CardTitle>
      <CardBody className="ocm--cost-breakdown-card__body">
        {!hasSources && sources.fulfilled ? (
          <CostEmptyState />
        ) : (
          <CostBreakdownSummary report={report as any as Report} />
        )}
      </CardBody>
    </Card>
  );
};

export default CostBreakdownCard;
