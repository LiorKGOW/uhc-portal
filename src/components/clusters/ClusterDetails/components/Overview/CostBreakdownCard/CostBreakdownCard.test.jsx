import React from 'react';
import { shallow } from 'enzyme';
import CostBreakdownCard from './CostBreakdownCard';

const initialState = {
  report: {
    data: [],
    meta: {},
    pending: false,
    fulfilled: true,
  },
  sources: {
    meta: {
      count: 0,
    },
    pending: false,
    fulfilled: true,
  },
};

const availableState = {
  report: {
    data: [
      {
        clusters: [
          {
            values: [
              {
                cluster: 'a94ea9bc-9e4f-4b91-89c2-c7099ec08427',
                clusters: [
                  'OCP-OnPrem01',
                ],
                infrastructure: {
                  raw: {
                    value: 0,
                    units: 'USD',
                  },
                  markup: {
                    value: 0,
                    units: 'USD',
                  },
                  usage: {
                    value: 39787.5631884,
                    units: 'USD',
                  },
                  total: {
                    value: 39787.5631884,
                    units: 'USD',
                  },
                },
                supplementary: {
                  raw: {
                    value: 0,
                    units: 'USD',
                  },
                  markup: {
                    value: 0,
                    units: 'USD',
                  },
                  usage: {
                    value: 0,
                    units: 'USD',
                  },
                  total: {
                    value: 0,
                    units: 'USD',
                  },
                },
                cost: {
                  raw: {
                    value: 0,
                    units: 'USD',
                  },
                  markup: {
                    value: 0,
                    units: 'USD',
                  },
                  usage: {
                    value: 39787.5631884,
                    units: 'USD',
                  },
                  total: {
                    value: 39787.5631884,
                    units: 'USD',
                  },
                },
              },
            ],
          },
        ],
      },
    ],
    meta: {},
    pending: false,
    fulfilled: true,
  },
  sources: {
    meta: {
      count: 1,
    },
    pending: false,
    fulfilled: true,
  },
};

describe('<CostBreakdownCard />', () => {
  let wrapper;
  let getReport;
  let getSources;
  describe('When no source providers are available', () => {
    beforeAll(() => {
      getReport = jest.fn();
      getSources = jest.fn();
      wrapper = shallow(
        <CostBreakdownCard
          getReport={getReport}
          getSources={getSources}
          sources={initialState.sources}
          report={initialState.report}
        />,
      );
    });

    it('renders empty state', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('calls getSources on mount', () => {
      expect(getSources).toBeCalled();
    });
    it('calls getReport on mount', () => {
      expect(getSources).toBeCalled();
    });
  });

  describe('When cost report is available', () => {
    beforeAll(() => {
      getReport = jest.fn();
      getSources = jest.fn();
      wrapper = shallow(
        <CostBreakdownCard
          getReport={getReport}
          getSources={getSources}
          sources={availableState.sources}
          report={availableState.report}
        />,
      );
    });

    it('renders cluster costs', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('calls getSources on mount', () => {
      expect(getSources).toBeCalled();
    });
    it('calls getReport on mount', () => {
      expect(getReport).toBeCalled();
    });
  });
});

/**

 */
