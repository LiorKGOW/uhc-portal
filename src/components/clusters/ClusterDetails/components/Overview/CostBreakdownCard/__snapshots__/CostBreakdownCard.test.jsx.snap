// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CostBreakdownCard /> When cost report is available renders cluster costs 1`] = `
<Card
  className="ocm--cost-breakdown-card"
>
  <CardTitle>
    <Title
      headingLevel="h2"
      size="lg"
    >
      Cost breakdown
    </Title>
  </CardTitle>
  <CardBody
    className="ocm--cost-breakdown-card__body"
  >
    <CostBreakdownSummary
      report={
        Object {
          "data": Array [
            Object {
              "clusters": Array [
                Object {
                  "values": Array [
                    Object {
                      "cluster": "a94ea9bc-9e4f-4b91-89c2-c7099ec08427",
                      "clusters": Array [
                        "OCP-OnPrem01",
                      ],
                      "cost": Object {
                        "markup": Object {
                          "units": "USD",
                          "value": 0,
                        },
                        "raw": Object {
                          "units": "USD",
                          "value": 0,
                        },
                        "total": Object {
                          "units": "USD",
                          "value": 39787.5631884,
                        },
                        "usage": Object {
                          "units": "USD",
                          "value": 39787.5631884,
                        },
                      },
                      "infrastructure": Object {
                        "markup": Object {
                          "units": "USD",
                          "value": 0,
                        },
                        "raw": Object {
                          "units": "USD",
                          "value": 0,
                        },
                        "total": Object {
                          "units": "USD",
                          "value": 39787.5631884,
                        },
                        "usage": Object {
                          "units": "USD",
                          "value": 39787.5631884,
                        },
                      },
                      "supplementary": Object {
                        "markup": Object {
                          "units": "USD",
                          "value": 0,
                        },
                        "raw": Object {
                          "units": "USD",
                          "value": 0,
                        },
                        "total": Object {
                          "units": "USD",
                          "value": 0,
                        },
                        "usage": Object {
                          "units": "USD",
                          "value": 0,
                        },
                      },
                    },
                  ],
                },
              ],
            },
          ],
          "fulfilled": true,
          "meta": Object {},
          "pending": false,
        }
      }
    />
  </CardBody>
</Card>
`;

exports[`<CostBreakdownCard /> When no source providers are available renders empty state 1`] = `
<Card
  className="ocm--cost-breakdown-card"
>
  <CardTitle>
    <Title
      headingLevel="h2"
      size="lg"
    >
      Cost breakdown
    </Title>
  </CardTitle>
  <CardBody
    className="ocm--cost-breakdown-card__body"
  >
    <CostEmptyState />
  </CardBody>
</Card>
`;
